File: src/lib/api/teamProvider.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  TeamProviderRegistry,
  TeamProviderSyncSettings,
  TeamSyncApplySummary,
  TeamSyncStatus,
} from "@/types";

export const teamProviderApi = {
  async getSyncSettings(): Promise<TeamProviderSyncSettings | null> {
    return (await invoke(
      "get_team_sync_settings",
    )) as TeamProviderSyncSettings | null;
  },

  async saveSyncSettings(
    settings: TeamProviderSyncSettings | null,
  ): Promise<void> {
    await invoke("save_team_sync_settings", { settings });
  },

  async getSyncStatus(): Promise<TeamSyncStatus> {
    return (await invoke("get_team_sync_status")) as TeamSyncStatus;
  },

  async fetchRegistry(sourceUrl: string): Promise<TeamProviderRegistry> {
    return (await invoke("fetch_team_registry", {
      sourceUrl,
    })) as TeamProviderRegistry;
  },

  async applyRegistry(sourceUrl: string): Promise<TeamSyncApplySummary> {
    return (await invoke("apply_team_registry", {
      sourceUrl,
    })) as TeamSyncApplySummary;
  },

  async resolveConflict(input: {
    providerId: string;
    app: string;
    acceptTeam: boolean;
    sourceUrl: string;
  }): Promise<void> {
    await invoke("resolve_team_sync_conflict", {
      request: {
        providerId: input.providerId,
        app: input.app,
        acceptTeam: input.acceptTeam,
        sourceUrl: input.sourceUrl,
      },
    });
  },

  async cleanupRemoved(): Promise<number> {
    return (await invoke("cleanup_removed_team_providers")) as number;
  },
};

File: src-tauri/src/commands/team_provider.rs
use tauri::{AppHandle, Emitter, State};

use crate::services::team_provider::{TeamProviderRegistry, TeamProviderService};
use crate::settings::{TeamProviderSyncSettings, TeamSyncApplySummary, TeamSyncStatus};
use crate::store::AppState;

#[tauri::command]
pub fn get_team_sync_settings() -> Result<Option<TeamProviderSyncSettings>, String> {
    Ok(crate::services::team_provider::auto_sync::get_sync_settings())
}

#[tauri::command]
pub fn save_team_sync_settings(settings: Option<TeamProviderSyncSettings>) -> Result<(), String> {
    crate::services::team_provider::auto_sync::save_sync_settings(settings)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_team_sync_status() -> Result<TeamSyncStatus, String> {
    Ok(crate::services::team_provider::auto_sync::get_sync_status())
}

#[tauri::command]
pub async fn fetch_team_registry(source_url: String) -> Result<TeamProviderRegistry, String> {
    TeamProviderService::fetch_registry_only(&source_url)
        .await
        .map(|(registry, _etag)| registry)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn apply_team_registry(
    app: AppHandle,
    state: State<'_, AppState>,
    source_url: String,
) -> Result<TeamSyncApplySummary, String> {
    let summary = TeamProviderService::apply_from_source(&state.db, &source_url)
        .await
        .map_err(|e| e.to_string())?;

    let _ = app.emit(
        "team-provider-synced",
        serde_json::json!({
            "source": "manual",
            "summary": summary,
        }),
    );

    Ok(summary)
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolveTeamSyncConflictRequest {
    pub provider_id: String,
    pub app: String,
    pub accept_team: bool,
    pub source_url: String,
}

#[tauri::command]
pub async fn resolve_team_sync_conflict(
    state: State<'_, AppState>,
    request: ResolveTeamSyncConflictRequest,
) -> Result<(), String> {
    let registry = TeamProviderService::fetch_registry_only(&request.source_url)
        .await
        .map(|(registry, _etag)| registry)
        .map_err(|e| e.to_string())?;

    TeamProviderService::resolve_conflict(
        &state.db,
        &request.provider_id,
        &request.app,
        request.accept_team,
        &registry,
        &request.source_url,
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn cleanup_removed_team_providers(state: State<'_, AppState>) -> Result<u32, String> {
    TeamProviderService::cleanup_removed(&state.db).map_err(|e| e.to_string())
}
